dex/elrond_dex_pair/src/oracle.rs

```diff
@@ -0,0 +1,257 @@
+    }
+
+    fn calculate_weighted_price(
+        &self,
+        weight_price: Self::BigUint,
+        weight_price_period: u64,
+        instant_price: Self::BigUint,
+        instant_price_period: u64,
+    ) -> Self::BigUint {
+        (weight_price * Self::BigUint::from(weight_price_period)
+            + instant_price * Self::BigUint::from(instant_price_period))
+            / Self::BigUint::from(weight_price_period + instant_price_period)
+    }
+
+    fn circular_binary_search(&self, block: Nonce) -> Option<PriceRecord<Self::BigUint>> {
+        let none = Option::None;
+        let mut low = 1;
+        let mut high = self.price_records().len();
+
+        if low > high {
+            return none;
+        }
+
+        while low <= high {
+            let mid = (low + high) / 2;
+            let mid_elem = self.price_records().get(mid);
+
+            if self.record_contains_block(&mid_elem, block) {
+                return Option::<PriceRecord<Self::BigUint>>::from(mid_elem);
+            }
+
+            let low_elem = self.price_records().get(low);
+            let high_elem = self.price_records().get(high);
+
+            #[allow(clippy::collapsible_else_if)]
+            if mid_elem.start_block <= high_elem.start_block {
+                if block > mid_elem.start_block && block <= high_elem.end_block {
+                    low = mid + 1;
+                } else {
+                    high = mid - 1;
+                }
+            } else {
+                if block >= low_elem.start_block && block < mid_elem.end_block {
+                    high = mid - 1;
+                } else {
+                    low = mid + 1;
+                }
+            }
+        }
+
+        none
+    }
+
+    fn record_contains_block(&self, record: &PriceRecord<Self::BigUint>, block: Nonce) -> bool {
+        record.start_block <= block && block <= record.end_block
+    }
+
+    #[view(getPriceRecordForBlock)]
+    fn get_price_record_for_block(&self, block: Nonce) -> Option<PriceRecord<Self::BigUint>> {
+        self.circular_binary_search(block)
+    }
+
+    #[view(getPriceRecordsBetweenRange)]
+    fn get_price_records_starting_between_range(
+        &self,
+        start: usize,
+        end: usize,
+    ) -> MultiResultVec<PriceRecord<Self::BigUint>> {
+        let mut result = MultiResultVec::<PriceRecord<Self::BigUint>>::new();
+        let default_value_fn = || PriceRecord::<Self::BigUint> {
+            first_token_price: Self::BigUint::zero(),
+            second_token_price: Self::BigUint::zero(),
+            start_block: 0u64,
+            end_block: 0u64,
+        };
+
+        let mut current_index = start;
+        while {
+            result.push(
+                self.price_records()
+                    .get_or_else(current_index, default_value_fn),
+            );
+            current_index = (current_index + 1) % RECORD_BUFFER_MAX_LEN;
+            current_index != end
+        } {}
```

Comment:
Rust has no such thing as _do while_, instead it has _do while do_ which can be used as a _do while_ if the second _do_ is an empty block. But still, replaced with loop syntax in order to avoid confusion